[WIP] Handle typing events from multiple users in a PM narrow

Subhasish-Behera opened the pull request:
<!-- Please see https://github.com/zulip/zulip-terminal#contributor-guidelines ! -->

**What does this PR do?**  <!-- Overall description goes here -->

<!-- If fixing a filed bug or new feature, add 'Fixes #<issue>' or 'Partial fix for #<issue>' -->
Fixes #992 
<!-- Add a link to a discussion on chat.zulip.org, if relevant -->

**Tested?** <!-- Fine to leave some of these unchecked if this is a draft/work-in-progress -->
- [X] Manually
- [X] Existing tests (adapted, if necessary)
- [X] New tests added (for any new behavior)
- [X] Passed linting & tests (each commit)
<!-- Code must pass CI (GitHub Actions) before merging - look for the green tick! -->

<!-- See https://github.com/zulip/zulip-terminal#commit-style -->
**Commit flow** <!-- if more than one commit; add/delete/fill-in as appropriate -->
<!-- For example:
- first commit doing some thing
- maybe multiple commits doing similar things
-->

**Notes & Questions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- For example:
- this doesn't include feature X (yet?)
- unsure about Y
- should this do Z?
-->
Basically now the `controller.active_conversation_info` is a `set` which adds a sender if the event is "start" and removes if the event is "stop". 
Reasons for change :-
1. There seems to be no requirement for Dict as It had only one key -"sender"
2. in dict , discarding/removing by element name is not possible (but possible in `set`)

**Interactions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- eg.
- Waiting on #<PR>
- Blocks #<PR>
-->

**Visual changes** <!-- if any; add/delete/fill-in with screenshot/diagram as appropriate -->

Subhasish-Behera commented on tests/core/test_core.py:
```diff
@@ -553,22 +553,22 @@ def test_maximum_popup_dimensions(
 
         def mock_typing() -> None:
-            controller.active_conversation_info = {}
+            controller.active_conversation_info = set()
```
In Line 571 and Line 589 ,The `{}` works for test whereas `set()` works for Lint

Subhasish-Behera commented on tests/model/test_model.py:
```diff
@@ -2882,7 +2882,7 @@ def test_toggle_stream_visual_notifications(
                 },
                 False,
-                {"sender_name": "hamlet"},
+                {"hamlet"},
```
In the cases of 1. `in_pm_narrow_with_sender_typing:start`, 
2. `in_pm_narrow_with_sender_typing:start_while_animation_in_progress`,
3 `in_pm_narrow_with_sender_typing:stop`,
when the test function calls `model._handle_typing_event(event)`, there is a AttributeError where `self.controller.active_conversation_info` is considered as `Dict` where as it is actually `Set`

neiljp commented on zulipterminal/model.py:
```diff
@@ -1374,7 +1374,7 @@ def _handle_typing_event(self, event: Event) -> None:
 
         narrow = self.narrow
         controller = self.controller
-        active_conversation_info = controller.active_conversation_info
+        active_conversation_info = self.controller.active_conversation_info
```
It's unclear to me why this change is necessary. The controller local variable is defined above this.

Subhasish-Behera replied:
My bad . I thought one way to clear the attribute error(in tests) could be to directly assign it (using self.controller.active_convesation_info rather than passing it from a variable). But that didn't work. So, this change is not necessary and I will fix it.

neiljp commented on zulipterminal/core.py:
```diff
@@ -79,7 +79,7 @@ def __init__(
         self._editor: Optional[Any] = None
 
-        self.active_conversation_info: Dict[str, Any] = {}
+        self.active_conversation_info: Set[Any] = set()
```
Whatever data structure we use here, we should move to a situation where we are not using `Any`, but rather the specific type being used.

Subhasish-Behera replied:
Will keep this in mind and change `Any` to `str`

neiljp commented on zulipterminal/core.py:
```diff
@@ -433,13 +433,31 @@ def show_typing_notification(self) -> None:
-                    ("footer_contrast", " " + sender_name + " "),
-                    " is typing" + next(dots),
-                ]
+            active_conversation_info = ", ".join(
+                map(str, self.active_conversation_info)
```
If these are already strings - try changing that `Any` I mention elsewhere - then this is misleading, since we don't need to apply that mapping.

neiljp commented on zulipterminal/core.py:
```diff
@@ -433,13 +433,31 @@ def show_typing_notification(self) -> None:
+            active_conversation_info = ", ".join(
+                map(str, self.active_conversation_info)
             )
+            if len(self.active_conversation_info) == 1:
```
In the following if/elif chain, there is a lot of common structure which is repeated. Try setting the footer text once, with variables that are set in the if/elif chain.

The logic will likely read clearer if you also give a name to the `len` before that logic.

mounilKshah replied:
I think you can modify it even further. If you notice, only the string containing `active_conversation_info` and *is/are typing* changes across all conditions. So you can to set only those strings in the switch case, and while passing the text to `set_footer_text()`, you can provide remaining elements.

Subhasish-Behera replied:
Yep. can be better. This version should address your suggestions
```
        while self.active_conversation_info:
           active_conversation_info = ", ".join(
                map(lambda x: self.model.user_dict[x]["full_name"], self.active_conversation_info)
            )
            no_of_typing_users = len(self.active_conversation_info)
            if no_of_typing_users == 1:
                typing_phrase = " is typing"
            elif no_of_typing_users < 4:
                typing_phrase = " are typing"
            else:
                active_conversation_info = "Multiple people"
                typing_phrase = " are typing"
            typing_text = [("footer_contrast",active_conversation_info+" ")]
            typing_text.append(typing_phrase+next(dots))
            self.view.set_footer_text(typing_text)
```
the whole initial active_conversation_info part just addresses Neil's indexing through the user_id(rather than full name) suggestion.

neiljp commented on zulipterminal/model.py:
```diff
@@ -1383,18 +1383,19 @@ def _handle_typing_event(self, event: Event) -> None:
             len(narrow) == 1
             and narrow[0][0] == "pm_with"
-            and sender_email in narrow[0][1].split(",")
+            and sender_email in [email.strip() for email in narrow[0][1].split(",")]
```
Is this change a bugfix? Why are we stripping?

Subhasish-Behera replied:
Yep. Earlier only the first user was detected as there was a space after the comma. This problem was mentioned in the issue as well ("In a huddle/group PM, we only handle typing events from the first email in the narrow specification.")

neiljp reviewed (commented):
@Subhasish-Behera Thanks for working on this :+1:

I've left a little feedback, but I'll follow up in the stream regarding the datastructure point (dict vs set).

mounilKshah reviewed (commented):
I tried running this PR on my local machine, and it worked well for PMs having 3 users and 2 typing simultaneously.
I had one query. All the tests currently present seem to have only one other user in the `pm_with` narrow, so test cases for multiple users typing simultaneously are yet to be implemented right?

Subhasish-Behera commented:
@zulipbot add "PR needs review"

theViz343 committed:
model/views: Fetch next unread topic using current message state.

This commit changes the behavior of the get_next_unread_topic method to
use the current message state (ie. the current topic) in calculating the
next unread topic to be cycled to.

If there is no current message available, ie. an empty narrow with no
focus, the logic attempts to use the current narrow to determine the
best course of action.

This replaces the previous approach of using a _last_unread_topic stored
in the model.

Tests updated.

Commit:
ui/buttons: On exiting topics view, associate topic name with stream_id.

This is achieved through a new method associate_stream_with_topic in the
View, which saves to a new internal dict.

This will support later restoring the topic position in the UI.

Since the topic list may be reordered between saving and restoring the
state, the current index in the topic list is insufficient; the name of
the topic is used intead.

theViz343 committed:
model: Add method to get stream topic from message id.

This commit adds a stream_topic_from_message_id method which returns the
topic of the currently focused message. This is done in preparation of
the change in algorithm of the get_next_unread_topic function.

Tests added.

theViz343 committed:
model: Change next_unread_topic to return None if topic stays same.

The next_unread_topic_from_message_id function returns the next topic
to be narrowed to. If the topic remains same, there is no need to call
narrow to the same topic again. This commit fixes this, without any
change in user-facing behavior.

Test case updated.

theViz343 committed:
refactor: model/views: Rename "get_next_unread_topic" method.

This commit renames the "get_next_unread_topic" function to
"next_unread_topic_from_message_id". This makes the function name
clearer by indicating that the function argument is the context from
which we get the next unread topic.

Tests updated to rename the function.

jagadeesh-akshaya committed:
keys/views: Add `=` key to toggle agreement with first message reaction.

Hotkeys document updated.

supascooopa committed:
messages: Add stream access type markers to stream message headers.

This improves the message UI by adding stream access type markers as
prefixes to stream names, when rendering headers of stream messages in
the messages view. This gives users an easier way to determine what type
of stream the message belongs to.

Tests and fixtures are updated accordingly.

neiljp commented on tests/core/test_core.py:
```diff
@@ -580,10 +580,30 @@ def mock_typing() -> None:
+                            ("footer_contrast", "hamlet "),
+                            ("footer", " is typing"),
+                        ]
+                    ),
```
I'm a bit confused why this reformatting was necessary? If you drop the comma after the second list element, does black shrink it back again?

neiljp commented on zulipterminal/core.py:
```diff
@@ -439,12 +439,17 @@ def show_typing_notification(self) -> None:
+            else:
+                active_conversation_info = "Multiple people"
+                typing_phrase = " are typing"
+            typing_text = [("footer_contrast", active_conversation_info + " ")]
+            typing_text.append(("footer", typing_phrase + next(dots)))
```
This part is definitely better with less duplication :+1:

neiljp commented on zulipterminal/model.py:
```diff
@@ -1392,7 +1392,8 @@ def _handle_typing_event(self, event: Event) -> None:
             elif event["op"] == "stop":
-                controller.active_conversation_info = {}
+                sender_email = self.user_dict[sender_email]["email"]
+                active_conversation_info.discard(sender_email)
```
This commit doesn't yet add tests for this case, or the output the user sees.

Does the model change along show reasonable improvements? If so, you could add a test change for the model in this commit, and move the core changes into the next commit with the test you added.

neiljp commented on tests/core/test_core.py:
```diff
@@ -578,35 +599,94 @@ def mock_typing() -> None:
+                        ),
+                    ]
+                )
+            elif length < 4:
```
This conditional should not be present in the test. Instead add a different parameter to the test for what you expect in each case.

neiljp commented on tests/model/test_model.py:
```diff
@@ -3021,10 +3030,131 @@ def test_toggle_stream_visual_notifications(
+                {"claudius@zulip.com", "hamlet@zulip.com"},
+                {"claudius@zulip.com", "hamlet@zulip.com"},
+                False,
+                id="in_group_pm_narrow,usertyping:stop_" "while animation in progress",
```
These ids look confusing.

neiljp commented on tests/core/test_core.py:
```diff
@@ -551,22 +551,28 @@ def test_maximum_popup_dimensions(
         mocker: MockerFixture,
         controller: Controller,
-        active_conversation_info: Dict[str, str],
+        active_conversation_info: Set[Any],
```
Check the types across commits - whether we settle on Set[str] as now or Set[int] for user ids, we shouldn't need to use Any now.

neiljp reviewed (commented):
@subhasish-Behera This looks to be working fine from a manual test :+1:

I left feedback inline. My main thought at this point, having not looked at the topic recently until again now just now, is that the logic would certainly be simpler if we tracked user ids instead of emails. We then have user_name_from_id to avoid diving into the internal model data, which should minimize some of the mocking.

This PR also currently has the commit text directly under the title as you had before, but that should be straightforward since you fixed other PRs already.

prah23 reviewed (changes requested):
Thanks for the updates @Subhasish-Behera!

My feedback would be to let `active_conversation_info` remain a `Dict` with emails as the keys and full names as the values. This way, the lookup for names in the controller's `show_typing_notifications` can be pushed to the model and can be performed in place of the line I marked in the comment below.

Subhasish-Behera commented on zulipterminal/model.py:
```diff
@@ -1381,18 +1381,19 @@ def _handle_typing_event(self, event: Event) -> None:
             if event["op"] == "start":
-                sender_name = self.user_dict[sender_email]["full_name"]
-                active_conversation_info["sender_name"] = sender_name
+                sender_email = self.user_dict[sender_email]["email"]
```
Yeah agreed. When I wrote the next line `active_conversation_info.add(sender_email)`, I was thinking that sender_email is the variable that we are talking about now to remove and not the previously declared `sender_email = event["sender"]["email"]`.
I think this can also be done in the `event["op"] == "stop"`

theViz343 committed:
model: Add in-stream wrap-around behavior to next unread topic behavior.

This commit aims to introduce in-stream wrap-around behavior to the
next_unread_topic_from_message_id function if there are unread messages
still present in the current stream.

Test case added.

neiljp committed:
pyproject: Expand and reorder mypy options to match mypy --help.

The strict section now matches entries and ordering from mypy --strict.

Other general checking options are extracted into their own sections
according to the order in the mypy --help page, except for those already
in the strict section.

Where the main command-line flag loosens the type-checking, these
options are treated as mypy defaults, and commented in each section.

This commit also starts enforcing the following, which require no
changes to code:
- strict_concatenate [strict]
- truthy-iterable [error-code]
- ignore-without-code [error-code]
- unused-awaitable [error-code]

neiljp committed:
helper/core: Remove unreachable code.

Confirmed after identification from mypy with warn_unreachable option.

neiljp committed:
refactor: api_types/views/tests: Correct types so mypy reaches code.

These changes ensure that mypy will not flag branches of code as
unreachable.

One type could be inferred instead, but is amended and remains to help
with readability, in _fetch_user_data.

neiljp committed:
pyproject: Enable mypy warn_unreachable checks.

This enforces this check after adjustments in previous commits.

This is a warning not included in mypy strict mode, so is added to a
new section.

neiljp committed:
run-mypy: Remove many hard-coded and command-line options.

These hard-coded options are now better stored in the pyproject.toml
configuration, while command-line options have not changed in some time
and are no longer necessary.

neiljp committed:
lister/run-mypy/pyproject: Improve follow-imports & type-check lister.

lister.py was previously explicitly excluded from type-checking, but if
we remove the hard-coded `follow-imports=silent` in tools/run-mypy, it
is evident that it was previously being checked due to being imported
where it was used - but error messages were not reported.

With minimal typing and a minor change to lister.py, the file can be
made to pass type-checking, enabling an overall move to the better
default `follow-imports=normal` setting in pyproject.toml. Moreover,
this also allows lister.py to be removed from the list of files excluded
from type-checking we maintain in pyproject.toml and run-mypy.

neiljp committed:
requirements[dev]: Add types-beautifulsoup4 to improve type-checking.

neiljp committed:
pyproject/tools: Error on missing types in libraries, with exceptions.

Rather than mypy simply ignoring any missing type information on
imports, this change sets this to be an error by default.

Separate sections for imports used in the main application and only
during development are added, to exclude certain libraries (import
patterns) from resulting in errors, where the type information is not
available. This makes the lack of typing more explicit and limited.

In addition, an import in tools/convert-unicode-emoji-data is marked
with a type-ignore comment, since the imported file is only temporary.

neiljp committed:
bugfix: messages: Correct edge cases in use of bs4 via type-checking.

This was surfaced by the use of the type stubs for beautifulsoup4.

Notes on the changes:
- tag_attr.get("attribute", []) can return a str or List[str] at
  runtime, depending upon whether the tag can be multivalued. Using
  .get_attribute_list() instead makes this explicit, in the absence of
  type stubs for .get() overloaded on the name of the parameter.
- Similarly, "title" is a single-entry field, so it is more accurate to
  set the default to the empty string.
- Lastly, .find() only applies to a Tag, and can result in a Tag, but
  also potentially None (already checked) or a NavigableString.
  Based on the structure of the document the latter is unlikely, but
  this is now ensured by converting the not-None check to whether the
  result is explicitly a Tag.

neiljp committed:
pyproject/views: Enable mypy truthy-bool error code checks.

Specific lines involving ModListWalker are ignored, since this class has an
implementation of __len__ in the base class, but that base class is
identified as Any due to limited type information at this time.

neiljp committed:
pyproject/model: Enable mypy enable_redundant_expr error code checks.

An always-true condition in _handle_user_settings_event is moved from a
conditional into an assert to satisfy this check.

neiljp committed:
run-mypy: Add support for repo/project-specific mypy arguments.

In general this behavior is now handled in pyproject.toml instead, but
in certain cases this is not possible.

This approach enables options to be applied differently when running
mypy on repos/projects (folders) which interact, eg. import from each
other, and changing one folder of files would otherwise be necessary
*only* to satisfy mypy when running on the other.

neiljp committed:
pyproject/run-mypy/api_types: Enable mypy no_implicit_reexport option.

This is enabled primarily in pyproject.toml.

The flags used when applying mypy to tests/ (in run-mypy) are
specifically adjusted to avoid this option. This cannot be achieved by
applying an override to "tests.*" in pyproject.toml, since the error is
flagged as being in imported non-test files. The source could be
adjusted to take this into account, but it appears cleaner to import
from the file being tested.

Two names imported from the zulip library are explicitly exported from
api_types.py to enable their use elsewhere in the source, using __all__.
The alternative is to use the `from X import Y as Y` technique, but this
triggers ruff error PLC0414 and is verbose.

neiljp committed:
model/buttons/messages/views: Import Message directly from api_types.py.

Other than shortening the chain of imports, this makes it clearer where
Message is originally defined, and that it is considered part of the
API (api_types.py), not an internal structure (helper.py).

neiljp committed:
pyproject/lint-and-test: Enable mypy untyped_decorators checks.

This is part of the strict block of checks.

This change passes mypy in a local full development environment, but to
pass in GitHub Actions the test dependencies must now be added to the
type-checking dependencies.

supascooopa committed:
symbols/messages: Add direct message marker to direct message headers.

The previous commit added a prefix marker for headers of stream
messages; this commit does the same for headers of direct messages.

No symbol was previously defined for direct messages, so one is added,
and applied in a similar way as with stream messages.

Tests updated.

neiljp committed:
refactor: model: Define & type user data together in model initializer.

Some fields were previously only defined in Model.get_all_users, but it
is cleaner to initialize them to sane values prior to that method call.
Others were defined above that call, but not directly, making it less
clear that they were directly updated by it.

neiljp committed:
refactor: model: Rename Model.get_all_users to reflect functionality.

This method does currently return some data, but also updates other data
as a side-effect, and is now only used internally to synchronize users
data from 'initial data' from API register() calls, or subsequent
updates to that data from events.

The revised name of _update_users_data_from_initial_data reflects these
factors.

Tests updated.

neiljp committed:
refactor: helper/ui_mappings/model: Move Literal types to helper.

StreamAccessType and UserStatus are internal types used to key into
ui_mappings dicts.

Since they are not part of the UI or Zulip API directly, other files
seeking to import them would be better served importing from helper.py
instead of the UI-centric mappings file.

neiljp committed:
refactor: model: Avoid Model.users special treatment via return value.

_update_users_data_from_initial_data also updates other model
attributes, so it's clearer to not treat it differently.

Tests updated.

neiljp committed:
refactor: helper/model: Introduce MinimalUserData to type user data.

Fixture and tests updated.

Subhasish-Behera committed:
ui/views: On initializing TopicsView, set topic focus from saved state.

Information regarding the previous position in the topic list for the
matching stream is retrieved from the View and used to set the initial
focus.

A new accessor in the View, saved_topic_in_stream_id, returns the most
recent (saved) topic name. In the absence of a previous value this
returns None.

A new internal TopicsView helper method returns the focus (index) which
corresponds to the saved topic name for the related stream. If there is
no saved state or no matching topic name, it returns the top index (0),
which was the previous behavior. This is isolated as a method to
facilitate testing.

Combined, this restores any previous topic-name state by assigning to
the focus position.

Test added for the internal helper method
_focus_position_for_topic_name.

supascooopa committed:
symbols: Add symbols (markers) for use with common narrows.

The 'Mentioned messages' and 'Starred messages' symbols are set to ASCII
characters, similar to those in the Zulip web app.

The selected 'All messages' symbol is the closest match found after some
testing, and appears to be available fairly widely. An appropriate
descriptive comment is added, as with other non-ASCII symbols, for later
reference.

mounilKshah committed:
refactor: tests: buttons: Inline IDs and extract SERVER_URL.

This commit shifts test IDs for test__parse_narrow_link() to be inline
with the test cases, and extracts SERVER_URL into the test function
body.

mounilKshah committed:
buttons: Add support for old format for narrow links.

This commit provides support for narrow links in message content
containing 'subject' instead of 'topic', which may be present in
messages before server version 2.1.0.

Test cases added.

Fixes #1422.

neiljp committed:
refactor: tests: buttons: Separate stream & message id in parsing links.

Minor change to reduce confusion between values of different ids in test
cases and ensure they are kept distinct for testing purposes.

supascooopa committed:
buttons/ui_sizes: Add prefix symbols to narrowing buttons in top-left.

This commit adds prefix symbols to the main buttons in the top left
corner of the UI. This aims to make the main buttons stand out more and
approximate the designs used in the web app.

This change also leads to a cleaner design, since these top buttons are
now indented and aligned similarly to the panels beneath them, eg. the
list of streams.

The symbol used as a prefix in headers of direct messages is reused for
the 'Direct messages' button, with other buttons using symbols added in
the previous commit.

The left part of the UI is increased in width to accommodate the new
additions.

Note that the "title" style is used to make the icons bolder, though
this should be decoupled in future.

theViz343 committed:
model/helper: Use stream panel order in sort_unread_topics.

This commit changes the sort_unread_topics method in helper.py
to use the left stream panel ordering to sort the unread data
instead of implicitly using the stream id as key.

Tests updated & extended.

neiljp committed:
requirements[dev]: Upgrade typos from ~=1.14.9 to 1.16.11.

This picks out O_WRONLY as a misspelling, but this version of typos
allows configuration in pyproject.toml, so exclude it there.

neiljp committed:
requirements[dev]: Upgrade codespell from ~=2.2.2 to ~=2.2.5.

neiljp committed:
pyproject/run-spellcheck: Recognize iterm & check docs/FAQ.md.

rsashank committed:
core: Prompt user before exiting using PopUpConfirmationView.

theViz343 committed:
refactor: model/helper: Extract unread_topics sorting into helper.py.

This commit introduces the sort_unread_topics function to the
next_unread_topic_from_message_id method to sort unread_topics
instead of sorting it in the model.

This replaces the use of bisect since the sorting behavior
changes due to the change in the sort_unread_topics function.
The caveat for this replacement is that the sort is performed
again.

This commit serves as a preparatory commit for the change in
sorting behavior in the next commit.

Test added.

Subhasish-Behera committed:
refactor: index active_conversation_info by id.

The active_converstation_information is a Dict where id of the users is
the key and the full_name of the user is the value.
Previously the indexing was done by name which is changed
to id now because of the possibility of non-unique names
present in the active_conversation_info.
In core.py active_conversation_info gets the full_names of the emails
 present in self.active_conversation_info's values.

Subhasish-Behera committed:
model/core: Add/Remove on start/stop event.

Add a single user and remove a single user on typing event.
Previously removing means emptying the dict as dict would contain
a single user.But now removing means discarding a specfic id.
In core.py diffrent cases of footer are made depending on no of active
user.
In test_core, the test now has it's own user_dict which is
used to extract name for footer notifcation out of emails
present in active_conversation_info.
In test_model,added a new parameter of current active_conversation_info.

neiljp committed:
requirements: Pin pygments at ~=2.15.1 instead of >=2.14.0.

Pygments 2.16.0 introduced a style to support a combination of bold and
italic styling in pygments/pygments#2444. Both of our gruvbox themes and
the light native theme gain a 'bold strong' style via pygments as a
result, which urwid fails to parse and blocks the application from
loading.

Longer-term we should improve the pygments to urwid translation logic to
allow these styles to work and an upgrade to later pygments versions,
but for now this allows these themes to continue working as before.

Fixes #1431.

Subhasish-Behera committed:
test_model: Added tests for _handle_typing_event_2.

Subhasish-Behera committed:
test_core: Add tests for show_typing_notification.

Add diffrernt cases of footer notification.Earlier
just one test was presnt as not much scenarios were
possible.
